amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+
+// Parts copied from failure:
+// https://github.com/rust-lang-nursery/failure
+
+#![warn(missing_docs, rust_2018_idioms, rust_2018_compatibility)]
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
+extern crate backtrace;
+
+use self::internal::Backtrace;
+use std::borrow::Cow;
+use std::error;
+use std::fmt;
+use std::result;
+
+/// Internal parts of `Error`.
+#[derive(Debug)]
+struct Inner<T: ?Sized> {
+    source: Option<Box<Error>>,
```

Comment:
Usually, after top-level error, the most interesting one is root cause.

Reply:
I would expect most error reporters to print the whole causal chain. Plus, errors are supposed to be "less frequent", so taking a performance hit when looking for the root cause doesn't seem overly severe.